### ISSUE TITLE
styled.* TS typings erase props of underlying JSX Element

### ISSUE THREAD

**Issue by eramdam** [callstack/linaria#456]
## Environment

Dependencies:

```
├── linaria@1.3.1
├── react@16.8.6
├── react-dom@16.8.6
└── typescript@3.5.2
```

## Description

Creating a styled component with `styled.*` returns a component whose props lack the ones from the underlying JSX element. Here's an example that repro

```js
import * as React from "react";
import { styled } from "linaria/react";
import { render } from "react-dom";

interface ButtonProps {
  isHovered?: boolean;
}

const StyledButton = styled.button<ButtonProps>`
  color: red;
`;

// Working but annoying work-around
const StyledButtonWorking = styled.button<
  ButtonProps & JSX.IntrinsicElements["button"]
>`
  color: red;
`;

export function MyComponent() {
  return <StyledButton className="" onClick={console.log} />;
}
```

Yields the following message

```
Type '{ className: string; onClick: { (message?: any, ...optionalParams: any[]): void; (message?: any, ...optionalParams: any[]): void; }; }' is not assignable to type 'IntrinsicAttributes & ButtonProps & { as?: "symbol" | "object" | "a" | "abbr" | "address" | "area" | "article" | "aside" | "audio" | "b" | "base" | "bdi" | "bdo" | "big" | "blockquote" | ... 161 more ... | undefined; } & { ...; }'.
  Property 'className' does not exist on type 'IntrinsicAttributes & ButtonProps & { as?: "symbol" | "object" | "a" | "abbr" | "address" | "area" | "article" | "aside" | "audio" | "b" | "base" | "bdi" | "bdo" | "big" | "blockquote" | ... 161 more ... | undefined; } & { ...; }'.
```

## Reproducible Demo

https://codesandbox.io/s/confident-sound-zw01e?from-embed

**Comment by nhooyr** [comment]
See https://github.com/callstack/linaria/pull/389

**Comment by eramdam** [comment]
@nhooyr Sweet! Any idea when this would land?

**Comment by brandonkal** [comment]
Fixed in my forked package, if you want something working right now.

**Comment by jayu** [comment]
solution merged #389
